rpcs3/Emu/Cell/PPUThread.cpp

```diff
@@ -485,19 +485,86 @@ extern void ppu_register_function_at(u32 addr, u32 size, u64 ptr)
 	return ppu_register_function_at(addr, size, reinterpret_cast<ppu_function_t>(ptr));
 }
 
+u32 ppu_get_exported_func_addr(u32 fnid, const std::string& module_name);
+
 struct ppu_far_jumps_t
 {
-	std::unordered_map<u32, std::pair<u32, bool>> vals;
+	struct all_info_t
+	{
+		u32 target;
+		bool link;
+		std::string module_name;
+		bool return_and_restore;
+		u64 saved_lr;
+		u64 saved_r2;
+	};
+
+	std::unordered_map<u32, all_info_t> vals;
 
 	mutable shared_mutex mutex;
 
-	std::pair<u32, bool> get_target(u32 pc) const
+	// Get target address, 'ppu' is used in ppu_far_jump in order to modify registers
+	u32 get_target(const u32 pc, ppu_thread* ppu = nullptr)
 	{
 		reader_lock lock(mutex);
 
 		if (auto it = vals.find(pc); it != vals.end())
 		{
-			return it->second;
+			all_info_t& all_info = it->second;
+			u32 target = all_info.target;
+
+			bool link = all_info.link;
+			bool from_opd = false;
+
+			if (!all_info.module_name.empty())
+			{
+				target = ppu_get_exported_func_addr(target, all_info.module_name);
+
+				// Avoid reading unmapped memory under mutex
+				if (vm::check_addr<sizeof(ppu_func_opd_t)>(target))
+				{
+					from_opd = true;
+				}
+			}
+
+			if (from_opd)
+			{
+				auto& opd = vm::_ref<ppu_func_opd_t>(target);
+				target = opd.addr;
+
+				// We modify LR to custom values here 
+				link = false;
+
+				if (ppu)
+				{
+					// Toggle flag
+					all_info.return_and_restore ^= true;
+
+					if (all_info.return_and_restore)
+					{
+						// Save LR and R2
```

Comment:
Aren't you supposed to save them in PPU context of current thread?

Reply:
Fixed.